### DISCUSSION TITLE
Node-canvas fonts

### DISCUSSION THREAD

**Discussion by Andrekarma** [fabricjs/fabric.js#10339] [General]
Hi,

i'm having a lot of trouble finding a way to obtain the exact same image result from a canvas in the browser and on server side. (node canvas version 2.10.2)

The main problem is with my custom fonts, that are correctly registered with node-canvas. 
But the final results has small differences in the fonts rendering from the browser. 

Anyone has faced this issue? Is there a solution/alternative to node-canva?
Any help would be great thanks! :)

![elemento_Andrea_test_lungo_proviamo12122024 (1) (1)](https://github.com/user-attachments/assets/f62d38b6-3507-4017-b9a5-894d99cc543e)

**Comment by asturur** [comment]
Font implementation is browser and OS dependant sadly.
1 - compare the output with firefox, it has the most similar rendering engine to node-canvas ( cairo + pango )
2 - are you rendering on node on the same machine or at least the same OS? difference in font configuration like font smoothing can have a role sometimes

I see the difference is mostly in spacing. Are you using letter spacing by chance?

**Comment by Andrekarma** [comment]
The machine is the same Windows machine and i am not using any letter spacing on those textbox.
There is a difference in the rendering between Chrome and Firefox, but the result with node-canvas still differs from both.

I think the problem relies mostly on Pango:
The image i sent has been done using node-canvas 2.10.2 with Pango 1.50.11
Moving to node-canvas 2.11.2 with Pango 1.50.14 i can see more accuracy on some fonts, but most of them are not rendered giving a "Pango-WARNING **: 11:56:04.851: couldn't load font "

So maybe with the latest version of pango there is a more faithful rendering, but this error on most fonts is preventing us from testing it (i did try to build the node-canvas library with pango 1.54.0)

**Reply by asturur** [reply]
Node canvas is shifting to pango 2 as far as i understand, hopefully that will solve the issues.
I can't do much to fix it as of now.
Something you could try is to add text-spacing as 0.001, that will force the library to render every character independently, and while slower by a lot, will defer the measurement of the distance of each char to javascript. Maybe will give you a different rendering that may be better for your use case

**Reply by Andrekarma** [reply]
Forcing the charSpacing to 0.0001 works very well, there are almost no differences between browser and node-canvas, but also between Chrome and Firefox.
Performance-wise ,as you said, it slowers down the image/pdf generation 
It is obviously a workaround for the moment, i hope node-canvas will fix it definitively.
But also for fabric i think it is important to have the same font rendering for all the browsers.

The line of code to force the font rendering (fabric 4.3.1)
fabric.Textbox.prototype.charSpacing = 0.0001;

Thanks @asturur (Grazie 😄 )

**Reply by asturur** [reply]
I do not know if from 4.3.1 there was any major text improvement that makes worth upgrading for your case.
I m not sure fabric can fix this generally since fabric does not produce text rendering but rather gives you an easy way to tell the browser what to render. Is just a different scope.

**Reply by Andrekarma** [reply]
@asturur i do have a strange effect when getting the pdf with node-canvas.
I am using the glyph mode canvas.lowerCanvasEl.getContext("2d").textDrawingMode = "glyph";
So the text inside the pdf is selectable

But if i set a color to the stroke on a textbox, the text is no more in glyph mode as you can see in the pdf attached (the text "Chiedi al banco")

[elemento_Farmacia_Creata27022025.pdf](https://github.com/user-attachments/files/19009953/elemento_Farmacia_Creata27022025.pdf)
 
Do you have any idea why?
Thanks

**Reply by asturur** [reply]
That chiedi al banco is clearly an image.
For some reason you are triggering the cache to be on.
It could be you are creating a group with a clippath or you are triggering a condition under the 'needsItsOwnCache'.

In the current version in this repo, search for `describe('needsItsOwnCache', () => {` and get an idea if that could be the case

**Reply by Andrekarma** [reply]
that's the problem
needsItsOwnCache: function() {
                return !("stroke" !== this.paintFirst || !this.hasFill() || !this.hasStroke() || "object" != typeof this.shadow) || !!this.clipPath
            },

paintFirst is "stroke" so the cache is used, it is fabric itself forcing that
Having the stroke set and forcing to not use the cache is going to remove the stroke

**Reply by asturur** [reply]
are you using shadow? There is a combination of effects that requires cache.
I have headache with all that negations there. That is an old version hopefully.
Also i don't see stroke on your pdf at all, what stroke are you using?

**Reply by asturur** [reply]
<img width="704" alt="image" src="https://github.com/user-attachments/assets/96abd5b0-bd6d-472a-9f26-28d8d413df43" />

is black text on grey button

**Reply by Andrekarma** [reply]
There is a black stroke on black text with strokeWidth = 1, so it is not really visible.
And no shadows at all
Here is more visible

[elemento_Genius_TV_Admin28022025.pdf](https://github.com/user-attachments/files/19026738/elemento_Genius_TV_Admin28022025.pdf)

**Reply by asturur** [reply]
if there is black stroke and black text, outside not being necessary, there could be just the issue that you have Shadow as null, that triggers the 'type object' just set the shadow as undefined and see if that was the issue.

If this is a commercial project consider upgrading too. I think you are o a very old version anyway

**Reply by Andrekarma** [reply]
Setting the shadow from null to undefined gives the text but removes the stroke
